lib/unread/readable.rb

```diff
@@ -27,11 +27,18 @@ def mark_array_as_read(array, reader)
             if global_timestamp && global_timestamp >= timestamp
               # The object is implicitly marked as read, so there is nothing to do
             else
-              rm = obj.read_marks.where(:reader_id => reader.id, :reader_type => reader.class.base_class.name).first || obj.read_marks.build
-              rm.reader_id   = reader.id
-              rm.reader_type = reader.class.base_class.name
-              rm.timestamp   = timestamp
-              rm.save!
+              # This transaction is needed, so that parent transaction won't rollback even there's an error.
+              ReadMark.transaction(requires_new: true) do
+                begin
+                  rm = obj.read_marks.where(reader_id: reader.id, reader_type: reader.class.base_class.name).first || obj.read_marks.build
+                  rm.reader_id   = reader.id
+                  rm.reader_type = reader.class.base_class.name
+                  rm.timestamp   = timestamp
+                  rm.save!
+                rescue ActiveRecord::RecordNotUnique
+                  raise ActiveRecord::Rollback
```

Comment:
In my tests with SQLite and MySQL removing the `raise ActiveRecord::Rollback` works fine. IMHO there is no need for a rollback here, because there is nothing to rollback. A Rollback is required if you want to undo a successful write operation, but we don't have one here, because the only write operation has already failed.

Some further investigation leads me to the following simplification of your change:

```ruby
else
  rm = obj.read_marks.where(:reader_id => reader.id, :reader_type => reader.class.base_class.name).first || obj.read_marks.build
  rm.reader_id   = reader.id
  rm.reader_type = reader.class.base_class.name
  rm.timestamp   = timestamp
  rm.save! rescue ActiveRecord::RecordNotUnique
```

No inner transaction needed, no thinking about savepoints, just add a rescue for the `save!` operation. What do you think?